### ISSUE TITLE
mido and python-midi parse some MIDI files differently

### ISSUE THREAD

**Issue by craffel** [olemb/mido#62]
Hi, I recently discovered this - I have found a few files where the package [`python-midi`](https://github.com/vishnubob/python-midi) parses a MIDI file differently than `mido`.  Here's a simple demonstration of that: 
```Python
>>> for t in mido.MidiFile('diff_mid1.mid').tracks:
...     print len(t)
...
4
1
1
7
342
1523
484
1660
682
3550
1094
1466
54

>>> for t in midi.read_midifile('diff_mid1.mid'):
...     print len(t)
...
6
7
342
1523
484
1660
682
3550
1094
1466
54
720
56
```
Not sure if you have any ideas as to why this is the case.  It seems pretty rare (found two files out of 1000 where this is true).  Here are the two example files: 
http://colinraffel.com/temp/diff_mid1.mid
http://colinraffel.com/temp/diff_mid2.mid

**Comment by olemb** [comment]
I've listed the messages for each track and found something interesting. Here are the first two tracks as returned by python-midi:

```
Track 1
midi.TimeSignatureEvent(tick=0, data=[4, 2, 24, 8])
midi.KeySignatureEvent(tick=0, data=[0, 0])
midi.SetTempoEvent(tick=0, data=[7, 144, 252])
midi.EndOfTrackEvent(tick=0, data=[])
midi.EndOfTrackEvent(tick=0, data=[])
midi.EndOfTrackEvent(tick=0, data=[])

Track 2
midi.TrackNameEvent(tick=0, text='Soft Karaoke', data=[83, 111, 102, 116, 32, 75, 97, 114, 97, 111, 107, 101])
midi.TextMetaEvent(tick=0, text='@KMIDI KARAOKE FILE', data=[64, 75, 77, 73, 68, 73, 32, 75, 65, 82, 65, 79, 75, 69, 32, 70, 73, 76, 69])
midi.TextMetaEvent(tick=0, text='@V0100', data=[64, 86, 48, 49, 48, 48])
midi.TextMetaEvent(tick=0, text='@I', data=[64, 73])
midi.TextMetaEvent(tick=0, text='@IRaffaele Sgueglia', data=[64, 73, 82, 97, 102, 102, 97, 101, 108, 101, 32, 83, 103, 117, 101, 103, 108, 105, 97])
midi.TextMetaEvent(tick=0, text='@Ie-mail: ssraf@mbox.vol.it', data=[64, 73, 101, 45, 109, 97, 105, 108, 58, 32, 115, 115, 114, 97, 102, 64, 109, 98, 111, 120, 46, 118, 111, 108, 46, 105, 116])
midi.EndOfTrackEvent(tick=0, data=[])
```

Here are the first four tracks as returned by Mido

```
Track 1
<meta message time_signature numerator=4 denominator=4 clocks_per_click=24 notated_32nd_notes_per_beat=8 time=0>
<meta message key_signature key='C' time=0>
<meta message set_tempo tempo=495868 time=0>
<meta message end_of_track time=0>

Track 2
<meta message end_of_track time=0>

Track 3
<meta message end_of_track time=0>

Track 4
<meta message track_name name='Soft Karaoke' time=0>
<meta message text text='@KMIDI KARAOKE FILE' time=0>
<meta message text text='@V0100' time=0>
<meta message text text='@I' time=0>
<meta message text text='@IRaffaele Sgueglia' time=0>
<meta message text text='@Ie-mail: ssraf@mbox.vol.it' time=0>
<meta message end_of_track time=0>
```

The 3 end of track messages seem to be messing things up. I'm not sure which library is reading this correctly. I'll have to investigate the file itself. There could also be other differences.

Thanks for reporting this!

**Comment by olemb** [comment]
Since it's risky to make any changes to the parser at this point I think it's time to add a comprehensive test suite. I have some files I've been testing with but I can't include them in the project and in any case it's best to have files that only contain the thing you're testing for.

**Comment by olemb** [comment]
I've got a fix here: e1b2f8f8e0b255c292cde967e6a09d0d9ecaf3d4
Both libraries now return the same tracks and messages.

I may have to make some changes to how `end_of_track` is handled elsewhere in the code. For example, if a track doesn't end with `end_of_track` the code in `_save()` will add one. Even though the MIDI file spec says that `end_of_track` is not optional, adding one may be a bad idea.

I've started working on the test suite. Let me know if you have any ideas for test cases.
